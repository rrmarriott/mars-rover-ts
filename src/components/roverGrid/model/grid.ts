export default class Grid {
    private _height: number;
    private _width: number;

    constructor(height: number, width: number) {
        this._height = height;
        this._width = width;
    }

    get height(): number {
        return this._height;
    }

    get width(): number {
        return this._width;
    }
}