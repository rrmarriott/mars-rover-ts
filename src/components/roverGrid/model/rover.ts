import Grid from './grid';
enum Command {
    MOVE_FORWARD,
    MOVE_BACKWARD,
    TURN_LEFT,
    TURN_RIGHT
}

enum RoverDirection {
    NORTH,
    SOUTH,
    EAST,
    WEST
}

class Rover {
    private static movement: Map<string, (rover: Rover) => Rover> = new Map<string, (rover: Rover) => Rover>();

    private _xPos: number;
    private _yPos: number;
    private _direction: RoverDirection;
    private _grid: Grid;

    static initMovements() {
        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.NORTH, Command.MOVE_FORWARD),
            (rover) => new Rover(rover.xPos, rover.yPos + 1, rover.direction, rover.grid));
        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.NORTH, Command.MOVE_BACKWARD),
            (rover) => new Rover(rover.xPos, rover.yPos - 1, rover.direction, rover.grid));

        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.SOUTH, Command.MOVE_FORWARD),
            (rover) => new Rover(rover.xPos, rover.yPos - 1, rover.direction, rover.grid));
        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.SOUTH, Command.MOVE_BACKWARD),
            (rover) => new Rover(rover.xPos, rover.yPos + 1, rover.direction, rover.grid));

        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.EAST, Command.MOVE_FORWARD),
            (rover) => new Rover(rover.xPos + 1, rover.yPos, rover.direction, rover.grid));
        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.EAST, Command.MOVE_BACKWARD),
            (rover) => new Rover(rover.xPos - 1, rover.yPos, rover.direction, rover.grid));

        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.WEST, Command.MOVE_FORWARD),
            (rover) => new Rover(rover.xPos - 1, rover.yPos, rover.direction, rover.grid));
        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.WEST, Command.MOVE_BACKWARD),
            (rover) => new Rover(rover.xPos + 1, rover.yPos, rover.direction, rover.grid));

        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.NORTH, Command.TURN_LEFT),
            (rover) => new Rover(rover.xPos, rover.yPos, RoverDirection.WEST, rover.grid));
        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.NORTH, Command.TURN_RIGHT),
            (rover) => new Rover(rover.xPos, rover.yPos, RoverDirection.EAST, rover.grid));

        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.SOUTH, Command.TURN_LEFT),
            (rover) => new Rover(rover.xPos, rover.yPos, RoverDirection.EAST, rover.grid));
        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.SOUTH, Command.TURN_RIGHT),
            (rover) => new Rover(rover.xPos, rover.yPos, RoverDirection.WEST, rover.grid));

        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.EAST, Command.TURN_LEFT),
            (rover) => new Rover(rover.xPos, rover.yPos, RoverDirection.NORTH, rover.grid));
        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.EAST, Command.TURN_RIGHT),
            (rover) => new Rover(rover.xPos, rover.yPos, RoverDirection.SOUTH, rover.grid));

        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.WEST, Command.TURN_LEFT),
            (rover) => new Rover(rover.xPos, rover.yPos, RoverDirection.SOUTH, rover.grid));
        Rover.movement.set(
            Rover.createMovementKey(RoverDirection.WEST, Command.TURN_RIGHT),
            (rover) => new Rover(rover.xPos, rover.yPos, RoverDirection.NORTH, rover.grid));
    }

    private static createMovementKey(direction: RoverDirection, command: Command) {
        return `${direction}-${command}`;
    }

    private static adjustPos(pos: number, size: number) {
        if (pos > size) {
            return 0;
        }

        if (pos < 0) {
            return size;
        }

        return pos;
    }

    constructor(xPos: number, yPos: number, direction: RoverDirection, grid: Grid) {
        this._xPos = Rover.adjustPos(xPos, grid.width - 1);
        this._yPos = Rover.adjustPos(yPos, grid.height - 1);
        this._direction = direction;
        this._grid = grid;
    }

    get xPos(): number {
        return this._xPos;
    }

    get yPos(): number {
        return this._yPos;
    }

    get direction(): RoverDirection {
        return this._direction;
    }

    get grid(): Grid {
        return this._grid;
    }

    move(commands: Array<Command>) {

        let newRover: Rover = this;

        commands.forEach(c => {
            let movementKey = Rover.createMovementKey(newRover._direction, c);
            let moveRover = Rover.movement.get(movementKey);

            newRover = moveRover ? moveRover(newRover) : newRover;
        });

        return newRover;
    }
}
Rover.initMovements();

export {RoverDirection, Rover, Command}