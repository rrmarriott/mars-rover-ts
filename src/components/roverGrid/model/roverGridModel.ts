import {viewBinding} from 'esp-js-react';
import RoverContainer from '../view/roverContainer';
import {Router, DisposableBase, observeEvent} from 'esp-js';
import {Rover, RoverDirection, Command} from './rover';
import Grid from './grid';

@viewBinding(RoverContainer)
export default class RoverGridModel extends DisposableBase {
    static readonly EXECUTE_COMMAND = 'executeCommand';
    static readonly MODEL_ID = 'marsRover';

    private _modelId: string;
    private _router: Router;
    private _rover: Rover;
    private _grid: Grid;

    constructor(router: Router) {
        super();
        this._modelId = RoverGridModel.MODEL_ID;
        this._router = router;

        this.createModelDefaults();
    }

    get modelId(): string {
        return this._modelId;
    }

    get rover(): Rover {
        return this._rover;
    }

    get grid(): Grid {
        return this._grid;
    }

    @observeEvent(RoverGridModel.EXECUTE_COMMAND)
    executeCommand(e: {command: Command}) {
        this._rover = this._rover.move([e.command]);
    }

    init() {
        this._router.addModel(this._modelId, this);
        this._router.observeEventsOn(this._modelId, this);
        this.addDisposable(() => {
            this._router.removeModel(this._modelId);
        });
    }

    private createModelDefaults() {
        this._grid = new Grid(20, 20);
        this._rover = new Rover(0, 0, RoverDirection.NORTH, this._grid);
    }
}