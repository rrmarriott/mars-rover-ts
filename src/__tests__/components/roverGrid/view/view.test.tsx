import * as renderer from 'react-test-renderer';
import * as React from 'react';
import RoverGrid from "../../../../components/roverGrid/view/roverGrid";
import {Rover, RoverDirection} from "../../../../components/roverGrid/model/rover";
import Grid from "../../../../components/roverGrid/model/grid";
import RoverGridCell from "../../../../components/roverGrid/view/roverGridCell";

test('Rover Grid renders correctly', () => {
    let grid = new Grid(5, 5);
    let rover = new Rover(0, 0, RoverDirection.NORTH, grid);

    const tree = renderer.create(
        <RoverGrid rover={rover} grid={grid} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});

test('Rover Grid Cell renders correctly', () => {
    let xPos = 0;
    let yPos = 0;
    let x = 5;
    let y = 5;

    const tree = renderer.create(
        <RoverGridCell xPos={xPos} yPos={yPos} x={x} y={y} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
});