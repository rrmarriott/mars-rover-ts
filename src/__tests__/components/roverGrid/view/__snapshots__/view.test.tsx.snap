exports[`test Rover Grid Cell renders correctly 1`] = `
<td
  className="RoverGridCell" />
`;

exports[`test Rover Grid renders correctly 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
      </tr>
      <tr>
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
      </tr>
      <tr>
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
      </tr>
      <tr>
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
      </tr>
      <tr>
        <td
          className="RoverGridCell RoverGridCellRover" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
        <td
          className="RoverGridCell" />
      </tr>
    </tbody>
  </table>
</div>
`;
